compiler/damlc/daml-doc/src/DA/Daml/Doc/Extract.hs

```diff
@@ -298,53 +267,91 @@ getFctDocs ctx@DocCtx{..} cl_nameM (DeclData decl docs) = do
         fct_anchor = Just $ functionAnchor dc_modname fct_name
         fct_descr = docs
 
-    guard $ case cl_nameM of
-        Just cl_name -> exportsField dc_exports cl_name fct_name
-        Nothing -> exportsFunction dc_exports fct_name
-
+    guard (exportsFunction dc_exports fct_name)
     Just FunctionDoc {..}
 
 getClsDocs :: DocCtx -> DeclData -> Maybe ClassDoc
-getClsDocs ctx@DocCtx{..} (DeclData (L _ (TyClD _ c@ClassDecl{..})) docs) = do
+getClsDocs ctx@DocCtx{..} (DeclData (L _ (TyClD _ ClassDecl{..})) tcdocs) = do
     let cl_name = Typename . packRdrName $ unLoc tcdLName
-        tyconMb = MS.lookup cl_name dc_tycons
-        cl_anchor = tyConAnchor ctx =<< tyconMb
-        cl_descr = docs
+    tycon <- MS.lookup cl_name dc_tycons
+    tycls <- tyConClass_maybe tycon
+    let cl_anchor = Just $ classAnchor dc_modname cl_name
+        cl_descr = tcdocs
         cl_args = map (tyVarText . unLoc) $ hsq_explicit tcdTyVars
-        cl_functions = map (dropMemberContext cl_anchor cl_args) $
-            concatMap (f cl_name) tcdSigs
+        opMap = MS.fromList
+            [ (Fieldname $ packId id, (id, dmInfoM))
+            | (id, dmInfoM) <- classOpItems tycls ]
+        cl_methods = concatMap (getMethodDocs opMap cl_anchor cl_name cl_args) subDecls
         cl_super = do
-            tycon <- tyconMb
-            cls <- tyConClass_maybe tycon
-            let theta = classSCTheta cls
+            let theta = classSCTheta tycls
             guard (notNull theta)
             Just (TypeTuple $ map (typeToType ctx) theta)
         cl_instances = Nothing -- filled out later in 'distributeInstanceDocs'
     guard (exportsType dc_exports cl_name)
     Just ClassDoc {..}
   where
-    f :: Typename -> LSig GhcPs -> [FunctionDoc]
-    f cl_name (L dloc (ClassOpSig p b names n)) = catMaybes
-      [ getFctDocs ctx (Just cl_name) (DeclData (L dloc (SigD noExt (ClassOpSig p b [L loc name] n))) (MS.lookup name subdocs))
-      | L loc name <- names
-      ]
-    f _ _ = []
+    -- All documentation of class members is stored in tcdDocs.
+    -- To associate the docs with the correct member, we convert all members
+    -- and docs to declarations, sort them by their location
+    -- and then use collectDocs.
+    -- This is the equivalent of Haddock’s Haddock.Interface.Create.classDecls.
+    subDecls :: [(LHsDecl GhcPs, Maybe DocText)]
```

Comment:
Note to myself and other reviewers: This corresponds to most of what `memberDocs` did before. What changed is that now we postprocess this with `getMethodDocs`.